### ISSUE TITLE
[Bug] RaggedPagedAttention v3 Kernel is slower for prefill-only workloads

### ISSUE THREAD

**Issue by alexshires** [sgl-project/sglang-jax#967]
### Checklist

- [x] 1. I have searched related issues but cannot get the expected help.
- [x] 2. The bug has not been fixed in the latest version.
- [x] 3. Please note that if the bug-related issue you submitted lacks corresponding environment info and a minimal reproducible demo, it will be challenging for us to reproduce and resolve the issue, reducing the likelihood of receiving feedback.
- [x] 4. If the issue you raised is not a bug but a question, please raise a discussion at https://github.com/sgl-project/sgl-jax/discussions/new/choose Otherwise, it will be closed.
- [x] 5. Please use English, otherwise it will be closed.

### Describe the bug

The sglang-jax RPA kernel v3 implementation is slower then RPA for prefill only workloads. The potential for this was already discussed here - https://github.com/sgl-project/sglang-jax/pull/913#issuecomment-4263329434 - but hopefully this is a clean performance test

v2 prefill average time: 0.001346 seconds
v3 prefill average time: 0.002209 seconds

v2 decode average time: 0.000294 seconds
v3 decode average time: 0.000352 seconds

v2 mixed average time: 0.001353 seconds
v3 mixed average time: 0.002225 seconds

### Reproduction

```
# Copyright 2026 Google LLC.
# SPDX-License-Identifier: Apache-2.0

import time
import jax
import jax.numpy as jnp
import pytest
from sgl_jax.srt.kernels.ragged_paged_attention.ragged_paged_attention import ragged_paged_attention as rpa_v2
from sgl_jax.srt.kernels.ragged_paged_attention.ragged_paged_attention_v3 import ragged_paged_attention as rpa_v3

def get_scenario_inputs(scenario, prompt_len=2000, num_q_heads=32, num_kv_heads=8, head_dim=128):
    """Helper to generate inputs based on the benchmark scenario."""
    batch_size = 1
    key = jax.random.PRNGKey(0)
    custom_mask = None
    
    if scenario == "prefill":
        q_len = prompt_len
        kv_lens = jnp.array([prompt_len], dtype=jnp.int32)
        page_indices = jnp.arange(prompt_len, dtype=jnp.int32)
        cu_q_lens = jnp.array([0, q_len], dtype=jnp.int32)
        cu_kv_lens = jnp.array([0, prompt_len], dtype=jnp.int32)
        distribution = jnp.array([0, batch_size, batch_size], dtype=jnp.int32)
        
    elif scenario == "decode":
        q_len = 1
        kv_lens = jnp.array([prompt_len], dtype=jnp.int32)
        page_indices = jnp.arange(prompt_len, dtype=jnp.int32)
        cu_q_lens = jnp.array([0, q_len], dtype=jnp.int32)
        cu_kv_lens = jnp.array([0, prompt_len], dtype=jnp.int32)
        distribution = jnp.array([batch_size, batch_size, batch_size], dtype=jnp.int32)
        
    elif scenario == "mixed":
        q_len = prompt_len
        kv_lens = jnp.array([prompt_len], dtype=jnp.int32)
        page_indices = jnp.arange(prompt_len, dtype=jnp.int32)
        cu_q_lens = jnp.array([0, q_len], dtype=jnp.int32)
        cu_kv_lens = jnp.array([0, prompt_len], dtype=jnp.int32)
        # Set [0, 0, 1] to target the MIXED branch of the kernel
        distribution = jnp.array([0, 0, batch_size], dtype=jnp.int32)
    else:
        raise ValueError(f"Unknown scenario: {scenario}")
        
    q = jax.random.normal(key, (q_len, num_q_heads, head_dim))
    k = jax.random.normal(key, (q_len, num_kv_heads, head_dim))
    v = jax.random.normal(key, (q_len, num_kv_heads, head_dim))
    
    return q, k, v, kv_lens, page_indices, cu_q_lens, cu_kv_lens, distribution, custom_mask

@pytest.mark.parametrize("rpa_version", [2, 3])
@pytest.mark.parametrize("scenario", ["prefill", "decode", "mixed"])
def test_rpa_performance(rpa_version, scenario):
    prompt_len = 2000
    num_kv_heads = 8
    head_dim = 128
    page_size = 1
    total_num_pages = 2000
    
    q, k, v, kv_lens, page_indices, cu_q_lens, cu_kv_lens, distribution, custom_mask = get_scenario_inputs(scenario, prompt_len)
    
    # Handle shape difference between v2 (4D) and v3 (5D)
    if rpa_version == 3:
        kv_cache_fused = jnp.zeros((total_num_pages, page_size, num_kv_heads * 2, 1, head_dim))
        kernel_fn = rpa_v3
    else:
        kv_cache_fused = jnp.zeros((total_num_pages, page_size, num_kv_heads * 2, head_dim))
        kernel_fn = rpa_v2
        
    # Warm up
    print(f"Warming up v{rpa_version} {scenario} kernel...")
    o, _ = kernel_fn(q, k, v, kv_cache_fused, kv_lens, page_indices, cu_q_lens, cu_kv_lens, distribution, custom_mask)
    jax.block_until_ready(o)
    
    num_iters = 20
    # Generate lists of inputs to avoid buffer donation errors
    q_list = [jax.random.normal(jax.random.PRNGKey(i), q.shape) for i in range(num_iters)]
    k_list = [jax.random.normal(jax.random.PRNGKey(i+100), k.shape) for i in range(num_iters)]
    v_list = [jax.random.normal(jax.random.PRNGKey(i+200), v.shape) for i in range(num_iters)]
    kv_cache_list = [jnp.zeros(kv_cache_fused.shape) for _ in range(num_iters)]
    
    print(f"Benchmarking v{rpa_version} {scenario} for {num_iters} iterations...")
    t0 = time.time()
    for i in range(num_iters):
        o, _ = kernel_fn(q_list[i], k_list[i], v_list[i], kv_cache_list[i], kv_lens, page_indices, cu_q_lens, cu_kv_lens, distribution, custom_mask)
        jax.block_until_ready(o)
    t1 = time.time()
    avg_time = (t1 - t0) / num_iters
    print(f"v{rpa_version} {scenario} average time: {avg_time:.6f} seconds")
    assert avg_time > 0
```

### Environment

python3 -m sgl_jax.check_env
Python: 3.12.13 (main, Apr 22 2026, 02:08:00) [GCC 14.2.0]
sglang-jax: 0.0.2
jax: 0.8.1
jaxlib: 0.8.1
triton: Module Not Found
transformers: 4.57.6
numpy: 2.2.6
aiohttp: 3.13.5
fastapi: 0.116.2
huggingface_hub: 0.34.6
modelscope: 1.28.2
orjson: 3.11.8
packaging: 26.1
psutil: 7.0.0
pydantic: 2.11.10
python-multipart: 0.0.26
pyzmq: 27.0.2
uvicorn: 0.35.0
uvloop: 0.21.0
openai: 2.32.0
tiktoken: 0.12.0
[TPU v6 lite-0]: TPU_0(process=0,(0,0,0,0))
ulimit soft: 1048576

**Comment by depksingh** [comment]
Hey @alexshires , I think there were some optimizations in Jax and libtpu latest versions which gave the v3 some better performance, the ideal versions would be Jax >= 0.8.2 and libtpu >= 0.0.38. Also I think the v3 RPA doesn't have tuned values yet for all configurations so you might need to tune it accordingly to get the best perf. @Iamleos I was working on the tuning script for v3 and is almost done, shall I create a PR for the same or are you already working on it?

Also, benchmarking the time with time.time() might not give the right metric for the latency, a better way would be to use sgl_jax/srt/kernels/utils/perf.py for the same.

**Comment by alexshires** [comment]
updated the script https://gist.github.com/alexshires/92f1e7061c90351a6222c1696609cc56 and with jax 0.8.2 (pinned to 0.0.32.1) I get

v2 prefill average time (device): 0.917316 milliseconds
v3 prefill average time (device): 0.995344 milliseconds
v2 decode average time (device): 0.323033 milliseconds
v3 decode average time (device): 0.333091 milliseconds
v2 mixed average time (device): 1.336017 milliseconds
v3 mixed average time (device): 2.201826 milliseconds

can you let me know when you have a tuning script?

**Comment by Iamleos** [comment]
We haven't tuned the best block size for different scenario
